File: companies/Amazon/02_LongestMountainInArray.java
class LongestMountainArray {
public static void main(String[] args)
{
    int []a = new int[args.length];
    for(int i=0; i<args.length; ++i) {    
        a[i] = Integer.parseInt(args[i]);
    }
    System.out.println(longestMountain(a));
}
    public static int longestMountain(int[] A) {
        int N   = A.length;
        int ans = 0, base = 0;
        
        while (base < N) {
            int end = base;
            
            // if base is a left-boundary
            if (end + 1 < N && A[end] < A[end + 1]) {
            
                // set end to the peak of this potential mountain
                while (end + 1 < N && A[end] < A[end + 1]) end++;

                // if end is really a peak..
                if (end + 1 < N && A[end] > A[end + 1]) {
                
                    // set end to the right-boundary of mountain
                    while (end + 1 < N && A[end] > A[end + 1]) end++;
                    
                    // record candidate answer
                    ans = Math.max(ans, end - base + 1);
                }
            }

            base = Math.max(end, base + 1);
        }

        return ans;
    }
}

File: companies/Amazon/02_LongestMountainInArray.cc
class Solution {
public:
    int longestMountain(vector<int>& arr) {
        
        int n = arr.size();
        if(n < 3) return 0;
        vector<int> pa(n, 1), qa(n, 1);
        
        for(int i=1; i<n; ++i){
            if(arr[i] > arr[i-1]){
                pa[i] += (pa[i-1] + 1);
            }
            if(arr[n-i-1] > arr[n-i]) {
                qa[n-i-1] += (qa[n-i]+1);
            }
        }
        
        int omax = 0;
        for(int i=0; i<n; ++i){
            if(pa[i] != 1 and qa[i] != 1)
                omax = max(omax, qa[i] + pa[i] - 1);
        }
        
        return (omax == 1 ? 0 : omax);
    }
};
